include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -397,115 +406,75 @@ struct is_vector<vector<T, N> > : bool_constant<true> {};
 
 #define NBL_REGISTER_OBJ_TYPE(T) NBL_REGISTER_TYPEID(T); \
-namespace nbl { namespace  hlsl { namespace type_traits {  namespace impl { \
-encoder<typeid_t<T>::value> encode_typeid(T); \
-template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> \
-{ using type = T; }; }}}}
+NBL_NAMESPACE_HLSL_TYPE_TRAITS_BEGIN namespace impl { \
+template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> { using type = T; }; \
+} NBL_NAMESPACE_HLSL_TYPE_TRAITS_END
 
 #define typeid(expr) (sizeof(::nbl::hlsl::type_traits::impl::encode_typeid(expr))/4)
 #define decltype(expr) ::nbl::hlsl::type_traits::impl::decltype_t<typeid(expr)>::type
 
 // builtins
-NBL_REGISTER_OBJ_TYPE(int16_t)
-NBL_REGISTER_OBJ_TYPE(int64_t)
-
-NBL_REGISTER_OBJ_TYPE(uint16_t)
-NBL_REGISTER_OBJ_TYPE(uint64_t)
-
-NBL_REGISTER_OBJ_TYPE(bool)
-NBL_REGISTER_OBJ_TYPE(bool4)
-NBL_REGISTER_OBJ_TYPE(bool3)
-NBL_REGISTER_OBJ_TYPE(bool2)
-NBL_REGISTER_OBJ_TYPE(bool1)
-
-NBL_REGISTER_OBJ_TYPE(int32_t)
-NBL_REGISTER_OBJ_TYPE(int32_t4)
-NBL_REGISTER_OBJ_TYPE(int32_t3)
-NBL_REGISTER_OBJ_TYPE(int32_t2)
-NBL_REGISTER_OBJ_TYPE(int32_t1)
-
-NBL_REGISTER_OBJ_TYPE(uint32_t)
-NBL_REGISTER_OBJ_TYPE(uint32_t4)
-NBL_REGISTER_OBJ_TYPE(uint32_t3)
-NBL_REGISTER_OBJ_TYPE(uint32_t2)
-NBL_REGISTER_OBJ_TYPE(uint32_t1)
-
-NBL_REGISTER_OBJ_TYPE(float32_t)
-NBL_REGISTER_OBJ_TYPE(float32_t4)
-NBL_REGISTER_OBJ_TYPE(float32_t3)
-NBL_REGISTER_OBJ_TYPE(float32_t2)
-NBL_REGISTER_OBJ_TYPE(float32_t1)
-
-NBL_REGISTER_OBJ_TYPE(float64_t)
-NBL_REGISTER_OBJ_TYPE(float64_t4)
-NBL_REGISTER_OBJ_TYPE(float64_t3)
-NBL_REGISTER_OBJ_TYPE(float64_t2)
-NBL_REGISTER_OBJ_TYPE(float64_t1)
-
-NBL_REGISTER_OBJ_TYPE(bool4x4)
-NBL_REGISTER_OBJ_TYPE(bool4x3)
-NBL_REGISTER_OBJ_TYPE(bool4x2)
-NBL_REGISTER_OBJ_TYPE(bool3x4)
-NBL_REGISTER_OBJ_TYPE(bool3x3)
-NBL_REGISTER_OBJ_TYPE(bool3x2)
-NBL_REGISTER_OBJ_TYPE(bool2x4)
-NBL_REGISTER_OBJ_TYPE(bool2x3)
-NBL_REGISTER_OBJ_TYPE(bool2x2)
-
-NBL_REGISTER_OBJ_TYPE(int32_t4x4)
-NBL_REGISTER_OBJ_TYPE(int32_t4x3)
-NBL_REGISTER_OBJ_TYPE(int32_t4x2)
-NBL_REGISTER_OBJ_TYPE(int32_t3x4)
-NBL_REGISTER_OBJ_TYPE(int32_t3x3)
-NBL_REGISTER_OBJ_TYPE(int32_t3x2)
-NBL_REGISTER_OBJ_TYPE(int32_t2x4)
-NBL_REGISTER_OBJ_TYPE(int32_t2x3)
-NBL_REGISTER_OBJ_TYPE(int32_t2x2)
-
-NBL_REGISTER_OBJ_TYPE(uint32_t4x4)
-NBL_REGISTER_OBJ_TYPE(uint32_t4x3)
-NBL_REGISTER_OBJ_TYPE(uint32_t4x2)
-NBL_REGISTER_OBJ_TYPE(uint32_t3x4)
-NBL_REGISTER_OBJ_TYPE(uint32_t3x3)
-NBL_REGISTER_OBJ_TYPE(uint32_t3x2)
-NBL_REGISTER_OBJ_TYPE(uint32_t2x4)
-NBL_REGISTER_OBJ_TYPE(uint32_t2x3)
-NBL_REGISTER_OBJ_TYPE(uint32_t2x2)
+
+#define NBL_REGISTER_MATRICIES(T) \
+    NBL_REGISTER_OBJ_TYPE(T) \
+    NBL_REGISTER_OBJ_TYPE(T ## x4) \
+    NBL_REGISTER_OBJ_TYPE(T ## x3) \
+    NBL_REGISTER_OBJ_TYPE(T ## x2) \
+
+#define NBL_REGISTER_TYPES_FOR_SCALAR(T) \
+    NBL_REGISTER_OBJ_TYPE(T) \
+    NBL_REGISTER_OBJ_TYPE(T ## 1) \
+    NBL_REGISTER_MATRICIES(T ## 2) \
+    NBL_REGISTER_MATRICIES(T ## 3) \
+    NBL_REGISTER_MATRICIES(T ## 4)
```

Comment:
you need to get into the habit of `#undef`-ing macros as soon as you're done using them

also you have a typo `MATRICES` not `MATRICIES`

Reply:
bad engrish